### DISCUSSION TITLE
bug: TRPC may create an invalid URL that crashes the server

### DISCUSSION THREAD

**Discussion by ivan-kleshnin** [trpc/trpc#6091] [General]
❗ Posting a bug here. I don't have enough time to prepare a sandbox, sorry. But it's almost certainly a bug and a bug in TRPC.

https://github.com/trpc/trpc/blob/next/packages/server/src/adapters/node-http/incomingMessageToRequest.ts#L70

```ts
const url = `http://${headers.get('host')}${req.url}`;
```

^ line assumes `url` contains a valid url. It's not always the case, as you can see, for example, in these tests:
https://github.com/sbordeyne/fastapi_spammer_protection/blob/fd4a3982c23cd4ab0f4761c9db603994c06fa19c/fastapi_spammer_protection/vulnerable_urls.py#L174

Some bots or spammers might send invalid requests with URLs like:
```
hotmail-com.olc.protection.outlook.com%3A25 (invalid!!!)
```

The above is invalid because it does not start with a slash. So `${headers.get('host')}${req.url}` becomes
something malformed of `{domain}{domain}` shape.

Here's how to emulate a request that crashes `adapters/node-http` (didn't test other adapters):
```
$ printf 'GET hotmail-com.olc.protection.outlook.com HTTP/1.1\r\nHost: localhost\r\n\r\n' | nc localhost 2000
```

Note that the server **crashes**, it does not just return 400 or 500 or something, but completely stops to accept new connections, basically goes offline❗ 

---

Solution: validate that `req.url` starts with `/`.

**Comment by KATT** [comment]
Please submit it as an issue even if you can't reproduce it, this seems urgent and we can validate it easily

**Reply by KATT** [reply]
@ivan-kleshnin have you actually seen cases of the url not starting with a `/` and it crashing an actual server?

**Reply by ivan-kleshnin** [reply]
Yes, it's not a theoretical guess. We've got some auxiliary servers, with exposed IPs, crashed by bots. Our main servers work behing Nginx which prevents the above from happening. So I've researched the issue down to this level, with time resources I currently have.

**Comment by ivan-kleshnin** [comment]
Move to https://github.com/trpc/trpc/issues/6094
